boot: Add printf functions

medhefgo opened the pull request:
This is the last big (and somewhat complicated) piece to add.

yuwata commented:
```
../src/boot/efi/util.c: In function ‘log_wait’:
/home/runner/work/systemd/systemd/src/fundamental/macro-fundamental.h:170:31: error: comparison of integer expressions of different signedness: ‘unsigned int’ and ‘int’ [-Werror=sign-compare]
  170 |                 UNIQ_T(A, aq) < UNIQ_T(B, bq) ? UNIQ_T(A, aq) : UNIQ_T(B, bq); \
      |                               ^
/home/runner/work/systemd/systemd/src/fundamental/macro-fundamental.h:165:19: note: in expansion of macro ‘__MIN’
  165 | #define MIN(a, b) __MIN(UNIQ, (a), UNIQ, (b))
      |                   ^~~~~
../src/boot/efi/util.c:419:19: note: in expansion of macro ‘MIN’
  419 |         BS->Stall(MIN(2 * log_count, 8) * 1000 * 1000);
      |                   ^~~
```

yuwata commented on src/boot/efi/util.c:
```diff
@@ -411,6 +411,15 @@ EFI_STATUS file_read(EFI_FILE *dir, const CHAR16 *name, UINTN off, UINTN size, C
+        if (log_count == 0)
+                return;
+
+        BS->Stall(MIN(2 * log_count, 8) * 1000 * 1000);
```
`8` -> `8u`??

yuwata commented:
semaphoreci
```
../src/boot/efi/efi-string.c: In function ‘format_status’: 01:14
../src/boot/efi/efi-string.c:367:14: error: ‘EFI_WARN_UNKNOWN_GLYPH’ undeclared (first use in this function); did you mean ‘EFI_WARN_UNKOWN_GLYPH’? 01:14
  367 |         case EFI_WARN_UNKNOWN_GLYPH:    s = u"Unknown glyph"; break; 01:14
      |              ^~~~~~~~~~~~~~~~~~~~~~ 01:14
      |              EFI_WARN_UNKOWN_GLYPH 01:14
../src/boot/efi/efi-string.c:367:14: note: each undeclared identifier is reported only once for each function it appears in
```

medhefgo commented:
Just FYI, github sends out notifications for CI failures on its own.

yuwata commented:
> Just FYI, github sends out notifications for CI failures on its own.

OK. Sorry for the noise.

medhefgo commented:
Some of the mkosi cis timeout out because sd-boot was failing. What concerns me is that this isn't detected as a ci failure.

@mrc0mmand Maybe you have some thoughts on that?

mrc0mmand commented:
> Some of the mkosi cis timeout out because sd-boot was failing. What concerns me is that this isn't detected as a ci failure.
> 
> @mrc0mmand Maybe you have some thoughts on that?

Eventually the job gets cancelled and the run with at least one cancelled job is marked as failed. However, the default timeout (360 minutes) is unnecessarily generous (imo), so maybe making it a bit shorter might help, e.g.:

```
diff --git a/.github/workflows/mkosi.yml b/.github/workflows/mkosi.yml
index 7061631c01..b0e120b3d3 100644
--- a/.github/workflows/mkosi.yml
+++ b/.github/workflows/mkosi.yml
@@ -25,6 +25,7 @@ env:
 jobs:
   ci:
     runs-on: ubuntu-latest
+    timeout-minutes: 60
     concurrency:
       group: ${{ github.workflow }}-${{ matrix.distro }}-${{ matrix.release }}-${{ github.ref }}
       cancel-in-progress: true
```

See: https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idtimeout-minutes

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+        if (hex_prefix) {
+                buf[0] = '0';
+                buf[1] = lower ? 'x' : 'X';
+                buf += 2;
```
```
*buf++ = '0';
*buf++ = lower ? 'x' : 'X';
```

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+
+        while (min_digits > n) {
+                buf[0] = '0';
+                buf++;
```
`*buf++ = '0';`

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+        }
+
+        while (n > 0) {
+                *buf = tmp[--n];
```
`*buf++ = tmp[--n];`

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+/* efi_vsnprintf is largely compatible to userspace vsnprintf. Any features omitted should trigger asserts.
+ * Notable differences:
+ *  - Passing NULL to %s is permitted and will print "(nil)".
+ *  - If buf is too small, -1 is returned instead of needed buf size (printf/xasprintf will do a allocation
```
"a allocation" -> "an allocation"

Maybe, "will do _for_ an allocation loop"??

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                                u = va_arg(*ap, int);
+                        else {
+                                if (!parse_number8(format, &u, &format) || u > INT_MAX)
+                                        assert_not_reached();
```
Even if the format is checked by the compiler with `_printf_` attribute, I'd like to return the function with `-EINVAL` and entirely refuse to formatting the remaining string.

medhefgo replied:
I would argue that any invalid format string goes to the same category as assert on internal function parameters. If we are fine with "don't call this function with a NULL pointer", it's only logical to apply the same for "don't call this with invalid/unsupported format string". All these are internal functions and aren't supposed to be exposed to outside users, so imho asserting should be safe and keeps things simpler.

Assert also makes it much more likely that an invalid/unsupported format does not get unnoticed instead of being silently swallowed somewhere.

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                }
+
+                case '%':
+                        *ret_str16 = u"%";
```
What happens when field width or precision is specified? E.g. `%9.9%`??

medhefgo replied:
gcc will complain about any of these: `%9.9% %9.% %0% %.0% %#%`

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+
+                        return format;
+
+                case 'm':
+                        format_number(tmp,
+                                      6,
+                                      /* hex= */ true,
+                                      /* has_prefix= */ true,
+                                      /* lower= */ true,
+                                      /* is_signed= */ false,
+                                      status);
```
Printing status as a number??

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                        format_number(tmp,
+                                      6,
+                                      /* hex= */ true,
+                                      /* has_prefix= */ true,
```
has_prefix -> hex_prefix?

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                                format_number(tmp,
+                                              /* min_digits= */ 8,
+                                              /* hex= */ true,
+                                              /* has_prefix= */ true,
```
Also here.

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+
+/* efi_vsnprintf is largely compatible to userspace vsnprintf. Any features omitted should trigger asserts.
+ * Notable differences:
+ *  - Passing NULL to %s is permitted and will print "(nil)".
```
I prefer "(null)" instead of "(nil)", as we use `strnull()` at many places.

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                const char16_t **ret_str16,
+                const wchar_t **ret_wstr) {
+
+        assert(ap);
```
Please add comments what kind of specifier is supported.

yuwata commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
 DEFINE_PARSE_NUMBER(char, parse_number8);
 DEFINE_PARSE_NUMBER(char16_t, parse_number16);
 
+static void format_number(
```
Maybe, it is better to introduce tests for this function.

yuwata replied:
Right. But, even so, still I'd like to have tests for the function, though that's not strong requirement.

yuwata reviewed (changes requested):
Reviewed only the first commit. Please rebase.

medhefgo commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,419 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                buf[0] = '-';
+                buf++;
+
+                int rem = -((int64_t) u % base);
+                u = (int64_t) u / -base;
+                tmp[n++] = digits[rem];
```
`INT64_MIN * -1` is an int overflow.

medhefgo commented:
I am hitting a really weird bug in some of the mkosi distros somewhere that causes the kernel to fail at some point during early boot.

The interesting part about this is that it seems to be caused by code size: it will break even when the printf functions aren't called from code whatsoever. The easiest way to trigger this was by changing the format_status lookup table. At some point I reduced code size to a single extra char in one of the strings. Interestingly, I cannot seem to reproduce this by only adding a very large string table onto the main branch (so without any of the printf functions defined).
Also note that despite the kernel breaking, the printf functions work perfectly fine.

In a dev branch where I have dropped libefi.a, this also doesn't happen. So my best guess is either some OVMF bug or the way the final binary is assembled (using gnu-efi).

Originally I simply decided to ignore this by dropping format_status until libefi.a is removed, but I found a simple workaround: building with at least -O1 seems to fix this (I even added a *giant* list of strings to the table just to make sure).

I have now looked at the functions for way longer than I like and cannot see anything that would cause this. I am also confused as to why other (mkosi) distros work fine. It doesn't seem to be because of distro-patched gnu-efi as at least debian seems to be using vanilla upstream.

I think using `-O1` until libefi.a is dropped is the way to go. I also reduced the size of this PR like I originally did to make it more palatable as it shouldn't be necessary to "rush" dropping libefi.a anymore.

poettering commented on src/boot/efi/assert.c:
```diff
@@ -6,7 +6,11 @@
 
 void efi_assert(const char *expr, const char *file, unsigned line, const char *function) {
-        log_error_stall(L"systemd-boot assertion '%a' failed at %a:%u, function %a(). Halting.", expr, file, line, function);
+        log_error("systemd-boot assertion '%s' failed at %s:%u, function %s(). Halting.",
```
actually "systemd-boot" is a misleading here, might also be in "systemd-stub"...

medhefgo replied:
This is an assert message, and changing this is too much work for little gain. Nobody is expected to see this. The important thing is the mention of systemd so that the right people get blamed. :D

poettering reviewed (changes requested):
looks pretty good to me. but i am of course not entirely sure i grapsed all details about the format string logic, but it does look ok, so let's hope the tests found the corner cases

medhefgo commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,500 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+        u"Buffer too small",
+        u"Stale data",
+        u"File system",
+        u"Reset required",
```
Imho, it's really "Not that many". I've changed it so that one of each table is provided for unit tests, though.

medhefgo commented:
I've simplified things quite a bit by moving the allocation loop into the actual printf functions. I also dropped snprintf as there's hardly any users for it.

I also decided to keep the differences to std printf to a minimum, so there will be no direct support for device paths, guids and optional printf output as I originally planned/did.

poettering commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
 
+enum {
+        SCRATCH_BUF_SIZE = 64,
+};
```
btw, I am not sure I agree with your love for enums even for stuff that is not actually enum-like or flags.

The thing is that in particuar for limits you typically want unsigned values actually, but enums are never unsigned. i.e. I think:

```c
#define SCRATCH_BUF_SIZE 64U
```

is much more appropriate actually.

medhefgo replied:
Don't bother looking at this code for now. I already majorly reworked this code to make it much nicer to look at. I am just doing some finishing touches right now.

poettering commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+
+                uint64_t rem = -((int64_t) u % base);
+                u = (int64_t) u / -base;
+                tmp[n++] = digits[rem];
```
(in userspace we usually use hexchar() for this)

poettering commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+        char16_t tmp[SCRATCH_BUF_SIZE];
+        const char *digits = FLAGS_SET(flags, NUMBER_LOWER) ? "0123456789abcdef" : "0123456789ABCDEF";
+
+        if (FLAGS_SET(flags, NUMBER_SIGNED) && (int64_t) u < 0) {
```
are you sure this cast is safe? isn't this one of these areas where the c compiler can do whatever it wants if u > INT64_MAX because undef behaviour or so?

medhefgo replied:
Yeah, it's implementation defined. And for gcc it's two's complement: https://gcc.gnu.org/onlinedocs/gcc/Integers-implementation.html#Integers-implementation
So this should be fine and testcases ought to be covering this anyways.

Also, it's interesting to know gcc only supports two's complement integers.

poettering commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
 DEFINE_PARSE_NUMBER(char16_t, parse_number16);
 
+enum {
+        SCRATCH_BUF_SIZE = 64,
```
can we determine how much precisely we need at max here? i.e. largest 64bit number in octal with most decoration we can have?

poettering commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+        for (;; format++)
+                switch (*format) {
+                case '0':
+                        *ret_pad_zero = true;
```
we have this coding style rule that on success we initialize all return params, and on failure none. afaics this might be set or might be left alone on success so this is conflicting with this rule.

You might be able to get away by simply dropping the "ret_" prefix, i.e. by stopping to claim that this is a return paremeter, but instead an input/output param.

poettering commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                        else {
+                                uint64_t u;
+                                if (!parse_number8(format, &u, &format) || u > INT_MAX)
+                                        assert_not_reached();
```
instead of:

```c
if (!x)
        asser_not_reached()
```

write:

```c
assert_se(x);
```

medhefgo replied:
This can never happen at runtime, as this is the passed inside of format string branch and will throw a compiler error:
`error: ‘%2147483648i’ directive output of 2147483648 bytes exceeds ‘INT_MAX’`

poettering commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                        return format;
+
+                case 'c':
+                        /* char/char16_t/wchat_t get promoted to int/wint_t. */
```
wchat is a typo

poettering commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                            status <= ((ELEMENTSOF(err_table) - 1) | EFI_ERROR_MASK)) {
+                                *ret_str8 = err_table[NOERR(status)];
+                                return format;
+                        }
```
i'd really add a helper call for this that takes the status as input and returns a const char * if it finds it in either of the two tables

poettering commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+        NUMBER_DEFAULT    = 1 << 0, /* Print as decimal. */
+        NUMBER_HEX        = 1 << 1, /* Print as hex. */
+        NUMBER_HEX_PREFIX = 1 << 2, /* Print as hex with 0x prefix. */
+        NUMBER_LOWER      = 1 << 3, /* Print hex numbers and prefix in lowercase letters. */
```
NUMBER_HEX_LOWER?

poettering commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                                break;
+                        case sizeof(int64_t):
+                                v = FLAGS_SET(flags, NUMBER_SIGNED) ? va_arg(*ap, int64_t) :
+                                                                      va_arg(*ap, uint64_t);
```
hmm, compilers don't warn that you return different types in both branches of the ternary op? weird...

medhefgo replied:
I think that only happens for pointers.

poettering commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+        if (ret_buf) {
+                if (buf == stack_buf) {
+                        *ret_buf = xnew(char16_t, n);
+                        memcpy(*ret_buf, buf, n * sizeof(*buf));
```
in userspace we have memdup() and newdup(), maybe add that here too?

medhefgo commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,453 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+                        char16_t *new_buf = xnew(char16_t, n_buf);
+                        memcpy(new_buf, buf, n * sizeof(*buf));
+
+                        free(dyn_buf);
```
That would free `stack_buf` for the first time the buffer is too small.

medhefgo commented:
I've restructured the code. It should be easier to grok now that there's less indentation and code that deals with adding to the buf moved into helper functions.

During all this I've found a ton of printf corner case behaviors that are all dealt with now (they should all be part of the unit tests). I even added support for more flags.

I am fairly confident about this implementation now.

keszybz commented:
Needs a rebase now (prolly because of the #24632 that was just merged).
Do we need to merge this before v252? It seems that there's some risk, even if it's a matter of issues somewhere in the toolchain. If possible, I'd prefer to merge this early in the next cycle.

medhefgo commented:
This can certainly wait for next cycle.

keszybz commented on src/boot/efi/secure-boot.c:
```diff
@@ -109,7 +109,7 @@ EFI_STATUS secure_boot_enroll_at(EFI_FILE *root_dir, const char16_t *path) {
                 err = efivar_set_raw(&sb_vars[i].vendor, sb_vars[i].name, sb_vars[i].buffer, sb_vars[i].size, sb_vars_opts);
                 if (err != EFI_SUCCESS) {
-                        log_error_stall(L"Failed to write %s secure boot variable: %r", sb_vars[i].name, err);
+                        log_error("Failed to write %ls secure boot variable: %m", sb_vars[i].name);
```
Hmm, `%m` should only be used with `log_*_errno`. `err` seems to have been lost.

medhefgo replied:
And here I thought I've stared enough times at these conversion to think I've fixed them all. :D

keszybz commented on src/boot/efi/secure-boot.c:
```diff
@@ -95,7 +95,7 @@ EFI_STATUS secure_boot_enroll_at(EFI_FILE *root_dir, const char16_t *path) {
                 err = file_read(dir, sb_vars[i].filename, 0, 0, &sb_vars[i].buffer, &sb_vars[i].size);
                 if (err != EFI_SUCCESS) {
-                        log_error_stall(L"Failed reading file %s\\%s: %r", path, sb_vars[i].filename, err);
+                        log_error("Failed reading file %ls\\%ls: %m", path, sb_vars[i].filename);
```
Here too.

keszybz commented on src/boot/efi/boot.c:
```diff
@@ -481,110 +468,140 @@ static void print_status(Config *config, char16_t *loaded_image_path) {
-                ps_string(L"        loader: %s\n", entry->loader);
+                printf("  config entry: %zu/%zu\n", i + 1, config->entry_count);
+                printf("            id: %ls\n", entry->id);
+                if (entry->title)
+                        printf("         title: %ls\n", entry->title);
```
This changes the logic to only print certain lines conditionally. Please add a comment why in the commit message.

keszybz commented on src/boot/efi/boot.c:
```diff
@@ -2031,10 +2048,9 @@ static EFI_STATUS boot_windows_bitlocker(void) {
 
         for (UINTN i = 0; i < boot_order_size / sizeof(uint16_t); i++) {
                 _cleanup_free_ char *buf = NULL;
-                char16_t name[sizeof(L"Boot0000")];
                 UINTN buf_size;
 
-                SPrint(name, sizeof(name), L"Boot%04x", (uint32_t) boot_order[i]);
+                _cleanup_free_ char16_t *name = xasprintf("Boot%04x", boot_order[i]);
```
Looks like this was supposed to be part of preceding commit.

keszybz reviewed (commented):
I didn't go into the details of the printf implementation, I was only mostly reviewing the usage of the new functions. This all looks pretty clean, apart from a few minor issues.

EDIT: I was reviewing the version before the last push. So some comments might be out-of-date.

medhefgo commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,495 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+        [EFI_WARN_DELETE_FAILURE]   = "Delete failure",
+        [EFI_WARN_WRITE_FAILURE]    = "Write failure",
+        [EFI_WARN_BUFFER_TOO_SMALL] = "Buffer too small",
+#  ifdef EFI_WARN_RESET_REQUIRED
+        [EFI_WARN_STALE_DATA]       = "Stale data",
+        [EFI_WARN_FILE_SYSTEM]      = "File system",
+        [EFI_WARN_RESET_REQUIRED]   = "Reset required",
+#  endif
```
gnu-efi doesn't provide these right now and I am too lazy to add them to missing_efi.h

medhefgo commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,495 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+        bool is_signed:1;
+        bool lowercase:1;
+        int8_t base;
+        char sign_pad; /* For + and (space) flags. */
```
Turns out that the struct is the same size regardless. Silly padding rules. :D

medhefgo commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,495 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+} SpecifierContext;
+
+typedef struct {
+        char16_t stack_buf[128]; /* We use stack_buf first to avoid allocations in most cases. */
```
This goes on the stack (and because it's char16_t it's double the bytes). Since this is efispace I'd rather be on the conservative side here. In practice it's only around one such allocation during a regular boot.

medhefgo commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,495 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+        ctx->n_buf = 2 * (ctx->n + need);
+
+        char16_t *new_buf = xnew(char16_t, ctx->n_buf);
+        memcpy(new_buf, ctx->buf, ctx->n * sizeof(*ctx->buf));
```
This could be stack_buf, which we shall remain forever imprisoned on stack. I've added a comment to make this clear.

keszybz replied:
Yeah, `stack_buf` would need to be handled explicitly.
```C
if (ctx->dyn_buf)
    realloc(ctx->dyn_buf, ctx->n * sizeof(*ctx->buf));
else {
    new_buf = xnew(char16_t, ctx->n_buf);
    memcpy(new_buf, ctx->buf, ctx->n * sizeof(*ctx->buf));
    free(ctx->dyn_buf);
}
```
After writing this out, I see that it's annoyingly longer. Let's keep the current version.

keszybz reviewed (commented):
LGTM.

We'll certainly want to create a fuzzer for the printf stuff. I can do it after this is merged, but if somebody else wants to, please go ahead.

yuwata commented:
needs to be rebased.

BTW, As I commented few weeks (months?) ago, this should be definitely tested with a fuzzer, though I do not block this PR without it, and it's ok to add that later. But, please please add a fuzzer for this feature.

medhefgo commented on src/boot/efi/efi-string.c:
```diff
@@ -272,6 +274,496 @@ bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
+        while (len > 0) {
+                len--;
+                ctx->buf[ctx->n++] = pad;
+        }
```
This should really be covered by the `assert(ctx.n < ctx.n_buf);` in `push_num/push_str` and the final one in `printf_internal`. (There is imho only so much paranoia before it gets silly.)

medhefgo commented:
How do you even fuzz printf? Random format strings itself would create tons of UB pitfalls thanks to the insanity that is vararg functions…

bluca commented:
needs a rebase

medhefgo commented:
I've added a fuzzer. Not sure how good that one is, though.

keszybz commented:
CI looks unhappy:

```
Run sudo timeout -k 30 10m mkosi qemu
Couldn't find swtpm binary, not invoking qemu with TPM2 device.
‣ Running command:
‣ qemu-system-x86_64 -machine type=q35,accel=tcg,smm=on -smp 1 -m 2G -object rng-random,filename=/dev/urandom,id=rng0 -device virtio-rng-pci,rng=rng0,id=rng-device0 -cpu max -nographic -nodefaults -serial mon:stdio -nic tap,script=no,downscript=no,ifname=vt-image,model=virtio-net-pci -drive if=pflash,format=raw,readonly=on,file=/usr/share/OVMF/OVMF_CODE.secboot.fd -global ICH9-LPC.disable_s3=1 -global driver=cfi.pflash01,property=secure,value=on -drive file=/var/tmp/.mkosi-0cazehvu,if=pflash,format=raw -drive 'if=none,id=hd,file=/home/runner/work/systemd/systemd/mkosi.output/centos_epel~8-stream/image.raw,format=raw' -device virtio-scsi-pci,id=scsi -device scsi-hd,drive=hd,bootindex=1

3hBdsDxe: loading Boot0001 "UEFI QEMU QEMU HARDDISK " from PciRoot(0x0)/Pci(0x3,0x0)/Scsi(0x0,0x0)
BdsDxe: starting Boot0001 "UEFI QEMU QEMU HARDDISK " from PciRoot(0x0)/Pci(0x3,0x0)/Scsi(0x0,0x0)
systemd-boot@0x7d356000,0x7d371000
!!!! X64 Exception Type - 0D(#GP - General Protection)  CPU Apic ID - 00000000 !!!!
ExceptionData - 0000000000000000
RIP  - 00000000709D43FC, CS  - 0000000000000038, RFLAGS - 0000000000000206
RAX  - 000A226DD2013274, RCX - 0000000000001380, RDX - 00000000709DC168
RBX  - 000000007DAB3B98, RSP - 000000007EFB4428, RBP - 000000007EFB47E0
RSI  - 00000000709DB0B8, RDI - 00000000709BC000
R8   - 0000000000000018, R9  - 0000000000000000, R10 - 000000007DAD1098
R11  - 000000007DC1B65C, R12 - 0000000000000000, R13 - 0000000000000000
R14  - 000000007EFB4568, R15 - 000000007DB1C298
DS   - 0000000000000030, ES  - 0000000000000030, FS  - 0000000000000030
GS   - 0000000000000030, SS  - 0000000000000030
CR0  - 0000000080010033, CR2 - 0000000000000000, CR3 - 000000007EC01000
CR4  - 0000000000000668, CR8 - 0000000000000000
DR0  - 0000000000000000, DR1 - 0000000000000000, DR2 - 0000000000000000
DR3  - 0000000000000000, DR6 - 00000000FFFF0FF0, DR7 - 0000000000000400
GDTR - 000000007E9E3000 0000000000000047, LDTR - 0000000000000000
IDTR - 000000007E4E3018 0000000000000FFF,   TR - 0000000000000000
FXSAVE_STATE - 000000007EFB4080
!!!! Find image based on IP(0x709D43FC) (No PDB)  (ImageBase=00000000709BC000, EntryPoint=00000000709C1000) !!!!
qemu-system-x86_64: terminating on signal 15 from pid 48414 (timeout)
Error: Process completed with exit code 124.
```

keszybz commented:
→ https://github.com/systemd/systemd/pull/25766

medhefgo commented:
It's just another case of overlapping PE sections (which in this case are not detected by the stub as code execution doesn't get far enough). The proper solution is porting mkosi to use ukify. Alternatively, one could use [this branch](https://github.com/medhefgo/systemd/tree/efi-clang-next), which does the little extra work to drop libefi.a, which should reduce the size pressure.

keszybz commented:
From the other PR:
```
 systemd-boot@0x7D0FB000,0x7D117000
systemd-stub@0x72349000,0x7235E000
Overlapping PE sections detected. Boot may fail due to image memory corruption!
systemd-boot assertion 'description' failed at ../src/src/boot/efi/measure.c:149, function tpm_log_event(). Halting.
```

Took me a while to see the issue:
```
Idx Name          Size      VMA               LMA               File off  Algn
  0 .text         00013aa0  0000000000005000  0000000000005000  00000370  2**4
                  CONTENTS, ALLOC, LOAD, READONLY, CODE
  1 .reloc        0000000a  0000000000019000  0000000000019000  00013f70  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  2 .data         000051a8  000000000001a000  000000000001a000  00014170  2**4
                  CONTENTS, ALLOC, LOAD, DATA
  3 .dynamic      00000100  0000000000020000  0000000000020000  00019370  2**2
                  CONTENTS, ALLOC, LOAD, DATA
  4 .osrel        0000029c  0000000000020000  0000000000020000  00019570  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  5 .rela         000014e8  0000000000021000  0000000000021000  00019970  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  6 .dynsym       00000018  0000000000023000  0000000000023000  0001af70  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  7 .sbat         000000d5  0000000000025980  0000000000025980  0001b170  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  8 .sdmagic      00000027  0000000000025a60  0000000000025a60  0001b370  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
  9 .cmdline      00000032  0000000000030000  0000000000030000  0001b570  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
 10 .linux        00c285e8  0000000002000000  0000000002000000  0001b770  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
 11 .initrd       038a76ee  0000000003000000  0000000003000000  00c43d70  2**2
                  CONTENTS, ALLOC, LOAD, READONLY, DATA
```
Both `.dynamic` and `.osrel` are allocated at the same VMA. This cannot work ;(

DaanDeMeyer commented:
@medhefgo Can you rebase this one? Should get rid of the mkosi CI failures

DaanDeMeyer commented:
@medhefgo needs another rebase it seems

medhefgo commented:
> @medhefgo needs another rebase it seems

I was just doing that when I saw https://github.com/systemd/systemd/pull/26087. This branch is just too damn easy to conflict with. :(

poettering commented on src/boot/efi/stub.c:
```diff
@@ -421,3 +416,15 @@ EFI_STATUS efi_main(EFI_HANDLE image, EFI_SYSTEM_TABLE *sys_table) {
+        EFI_STATUS err = real_main(image);
+        log_wait();
+        return err;
+}
```
this reminds me a lot of what we have in userspace with DEFINE_MAIN_FUNCTION(). I think we might want to do something similar here sooner or later?

(but i guess that can be done another time)

medhefgo replied:
I'll think on it

poettering commented on src/fundamental/macro-fundamental.h:
```diff
@@ -74,11 +74,11 @@
         #endif
         #define static_assert _Static_assert
-        #define assert_se(expr) ({ _likely_(expr) ? VOID_0 : efi_assert(#expr, __FILE__, __LINE__, __PRETTY_FUNCTION__); })
+        #define assert_se(expr) ({ _likely_(expr) ? VOID_0 : efi_assert(#expr, __FILE__, __LINE__, __func__); })
```
if you change this, please also change this for the other asserts in src/basic/macro.h

i.e. across the whole codebase please.

if i "git grep" through things right now I see most `__func__`, the asserts re the only holdouts.

poettering commented:
lgtm

medhefgo commented:
I converted the other `__PRETTY_FUNCTION__` too.
